tools/fast_build_tar/buildtar.go

```diff
@@ -0,0 +1,531 @@
+			break
+		}
+		if err != nil {
+			return err
+		}
+		header.Name = filepath.Join(root, header.Name)
+		if header.Typeflag == tar.TypeDir && !strings.HasSuffix(header.Name, "/") {
+			header.Name = header.Name + "/"
+		}
+		if err := f.tw.WriteHeader(header); err != nil {
+			return err
+		}
+		if _, err = io.Copy(f.tw, tr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *tarFile) addDeb(toAdd string) error {
+	return fmt.Errorf("addDeb unimplemented")
+}
+
+func (f *tarFile) makeDirs(header tar.Header) error {
+	dirToMake := []string{}
+	dir := header.Name
+	for {
+		dir = filepath.Dir(dir)
+		if dir == "." || dir == "/" {
+			break
+		}
+		dirToMake = append(dirToMake, dir)
+	}
+	for i := len(dirToMake) - 1; i >= 0; i-- {
+		dir := dirToMake[i]
+		if _, ok := f.dirsMade[dir]; ok {
+			continue
+		}
+		dh := header
+		dh.Typeflag = tar.TypeDir
+		dh.Name = dir + "/"
+		if err := f.tw.WriteHeader(&dh); err != nil {
+			return err
+		}
+
+		f.dirsMade[dir] = struct{}{}
+	}
+	return nil
+}
+
+func (f *tarFile) Close() {
+	for i := len(f.closers) - 1; i >= 0; i-- {
+		f.closers[i]()
+	}
+}
+
+func main() {
+	var (
+		flagfile string
+
+		output      string
+		directory   string
+		compression string
+
+		files multiString
+		tars  multiString
+		debs  multiString
+		links multiString
+
+		mode  string
+		modes multiString
+
+		owner      string
+		owners     multiString
+		ownerName  string
+		ownerNames multiString
+	)
+
+	flag.StringVar(&flagfile, "flagfile", "", "Path to flagfile")
+
+	flag.StringVar(&output, "output", "", "The output file, mandatory")
+	flag.StringVar(&directory, "directory", "", "Directory in which to store the file inside the layer")
+	flag.StringVar(&compression, "compression", "", "Compression (`gz` or `bz2`), default is none.")
+
+	flag.Var(&files, "file", "A file to add to the layer")
+	flag.Var(&tars, "tar", "A tar file to add to the layer")
+	flag.Var(&debs, "deb", "A debian package to add to the layer")
+	flag.Var(&links, "link", "Add a symlink a inside the layer ponting to b if a:b is specified")
+
+	flag.StringVar(&mode, "mode", "", "Force the mode on the added files (in octal).")
+	flag.Var(&modes, "modes", "Specific mode to apply to specific file (from the file argument), e.g., path/to/file=0455.")
+
+	flag.StringVar(&owner, "owner", "0.0", "Specify the numeric default owner of all files, e.g., 0.0")
+	flag.Var(&owners, "owners", "Specify the numeric owners of individual files, e.g. path/to/file=0.0.")
+	flag.StringVar(&ownerName, "owner_name", "", "Specify the owner name of all files, e.g. root.root.")
+	flag.Var(&ownerNames, "owner_names", "Specify the owner names of individual files, e.g. path/to/file=root.root.")
+
+	flag.Set("alsologtostderr", "true")
+
+	flag.Parse()
+
+	if flagfile != "" {
+		b, err := ioutil.ReadFile(flagfile)
+		if err != nil {
+			glog.Fatalf("couldn't read flagfile: %v", err)
+		}
+		cmdline := strings.Split(string(b), "\n")
+		flag.CommandLine.Parse(cmdline)
```

Comment:
does this mean that the flagfile overrides the cmdline args? is that what we want? (maybe we don't care?)

Reply:
It overrides non multistring args, yes. These is very close to python flagfile but I didn't want to rewrite the parser. It's not an exact match. gflag says:

> Flags are always processed in the expected order. That is, processing begins by examining the flags specified directly on the command line. If a flagfile is specified, its contents are processed, and then processing continues with remaining flags from the command line.